klippy/extras/filament_switch_sensor.py

```diff
@@ -0,0 +1,165 @@
+# Generic Filament Sensor Module
+#
+# Copyright (C) 2019  Eric Callahan <arksine.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import logging
+
+class BaseSensor(object):
+    def __init__(self, config):
+        self.name = config.get_name().split()[1]
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.runout_gcode = config.get('runout_gcode', None)
+        self.autoload_gcode = config.get('autoload_gcode', None)
+        self.runout_pause = config.getboolean('pause_on_runout', True)
+        if self.runout_pause:
+            if self.runout_gcode is None:
+                self.runout_gcode = "PAUSE"
+            else:
+                self.runout_gcode = "PAUSE\n" + self.runout_gcode
+        self.runout_cb = self._runout_event_handler
+        self.insert_cb = self._autoload_event_handler
+        self.runout_enabled = False
+        self.insert_enabled = False
+        self.event_running = False
+        self.monitor_state = True
+        self.print_status = "idle"
+        for status in ["idle", "ready", "printing"]:
+            self.printer.register_event_handler(
+                "idle_timeout:%s" % (status),
+                (lambda e, s=self, st=status: s._update_print_status(st)))
+        self.printer.register_event_handler("klippy:ready", self._handle_ready)
+        self.gcode.register_mux_command(
+            "QUERY_FILAMENT_SENSOR", "SENSOR", self.name,
+            self.cmd_QUERY_FILAMENT_SENSOR,
+            desc=self.cmd_QUERY_FILAMENT_SENSOR_help)
```

Comment:
Okay, I will make that change.  I do think it would still be wise to keep the following in the base class:
```python
def cmd_QUERY_FILAMENT_SENSOR(self, params):
        raise NotImplementedError(
            "Sensor must implement cmd_QUERY_FILAMENT_SENSOR")
```
This makes sure subclasses implement the correct gcode.